### ISSUE TITLE
WCAG 2.2 focus visible enhanced contrast seems to leave gap?

### ISSUE THREAD

**Issue by patrickhlauke** [w3c/wcag#1038]
I'm sure I queried this before, but can't seem to find it here on GitHub...

The proposed new Focus Visible (Enhanced) #936 / https://w3c.github.io/wcag/guidelines/22/#focus-visible-enhanced seems to, at least in my reading, leave an odd gap when it comes to contrast.

> Contrast or thickness: The focus indication area has a 3:1 contrast ratio against all adjacent colors for the minimum area or greater, ***or has a thickness of at least 2 CSS pixels***.
>
> NOTE: ***A focus indicator that is larger than the minimum area*** may have parts that do not meet the 3:1 contrast ratio, as long as an area equal to the minimum does meet the contrast ratio.

Assume you have a focus indication that conforms to the minimum area (and *not* any larger) and is 2 CSS pixels thick:

- "Contrast or thickness: ... *OR* has a thickness of at least 2 CSS pixels" is satisfied

- "Note: a focus indicator that is larger..." the entire note does not apply to this case, as it's *not* larger

So, if it's only minimum area sized, and 2 CSS pixels thick, it can have a contrast of 1.1:1?

If the idea is that no, it also still needs to pass 1.4.11 Non-text Contrast anyway...what's the point of all the above? Or alternatively, would it not need to have this specifically spelled out in the note ("in all cases, the minimum focus indication area must conform to 1.4.11 and have a contrast ratio of 3:1 against all adjacent colors"? or drop that "or has a thickness of at least 2 CSS pixels" escape clause?

**Comment by patrickhlauke** [comment]
Or does this need to be refactored altogether to say, in essence:

- the minimum area of the focus indicator must have a contrast of 3:1 in all cases
- anything than exceeds that minimum area can have whatever contrast you like

and forget about the 2 CSS px altogether? was that the intention?

**Comment by alastc** [comment]
Need to co-ordinate with #1003, also on the sizing metric.

**Comment by alastc** [comment]
The way it was intended was that the note refers to the contrast _change_ only.

I.e. 
> NOTE: A focus indicator that is larger than the minimum area may have parts that do not meet the 3:1 contrast ratio **change**, as long as an area equal to the minimum does meet the contrast ratio.

The thickness aspect is only for the adjacent contrast (3rd bullet) and does not combine with the change of contrast (1st bullet).

**Comment by patrickhlauke** [comment]
sorry, but...i would really suggest this gets unrolled and looked at again. it is incomprehensible to mere mortals. if you're differentiating between contrast change of all the bits from unfocused to focused (bullet 2?), and *then* contrast between the focus indicator(s) in the focused state and all adjacent other colors, then that needs to be made a lot clearer in the bolded bit / terms of the bullet points. this current wording is twisting my brain, and i'm not new to this topic...

**Comment by alastc** [comment]
(Noting the thumbs up from Jake, who is trying to add more complexity and bullets...?!)

Sorry, yes, 2nd bullet (updated).

In terms of process, I think we're best off continuing with the public review, gathering all feedback and then dealing with it (including Jakes, which potentially goes the other way).

**Comment by alastc** [comment]
Also, re-reading the top comment, the idea of the note was to account for gradients, where (like NTC) there can be areas of a thing (e.g. focus indictor) which do and don't meet it, so we're looking for a minumum size of that contrast level.

**Comment by jake-abma** [comment]
`(Noting the thumbs up from Jake, who is trying to add more complexity and bullets...?!)`

:-) IF you suggest to go into a certain direction and hold on to it, at least it should be done in a way covering the complete spectrum of scenarios. Or find another way.

**Comment by jake-abma** [comment]
Please also note that my suggestion was simple, a minimum surface of recognition, because that's what we want to achieve in the first place (people 'seeing' the focus). Not related to a certain calculation formula based on target size.

This was not honored by... and thus we're stuck with a calculation not clear and not covering all possible and logical scenarios.

We might want to re-consider?!

**Comment by patrickhlauke** [comment]
> In terms of process, I think we're best off continuing with the public review, gathering all feedback and then dealing with it (including Jakes, which potentially goes the other way).

consider this part of my public feedback. "this thing is pretty much incomprehensible as it stands"

**Comment by AutoSponge** [comment]
If you remove "**or** has a thickness of at least 2 CSS pixels" (bold is mine) it seems clear. By only changing a 2px border, you can pass all three rules as written. See: https://codepen.io/AutoSponge/pen/QWbyMrM?editors=1100 but that's hardly a visible focus indicator.

**Comment by alastc** [comment]
> If you remove "or has a thickness of at least 2 CSS pixels" (bold is mine) it seems clear.

So making it an 'and'? In which case all indicators have to be 2px thick _and_ contrast with both the previous color and adjacent color? 

That massilve increases the requirement and makes some fail which are very visible. It's worth having a look through lots of the test cases for this:
https://alastairc.uk/tests/wcag22-examples/focus-more-visible.html
https://alastairc.uk/tests/wcag22-examples/focus-more-visible-2.html
https://alastairc.uk/tests/wcag22-examples/focus-more-visible-3.html
https://alastairc.uk/tests/wcag22-examples/focus-more-visible-4.html

**Comment by patrickhlauke** [comment]
as suggested, if you're differentiating two different types of "contrast", then maybe making it clearer that this refers to different things being contrasted/measured would be a start

**Comment by alastc** [comment]
Maybe I'm in a woods for the trees situation, but I'm struggling to see how?

The two contrast bullets are:
> **Focus contrast:** Color changes used to indicate focus have at least a 3:1 contrast ratio with the colors changed from the unfocused control.
> **Contrast or thickness:** The focus indication area has a 3:1 contrast ratio against all adjacent colors for the minimum area or greater, or has a thickness of at least 2 CSS pixels.

One is for 'changed colors', the other for 'adjacent colors'. Looking for ideas to differentiate better...

**Comment by alastc** [comment]
[Branch created](https://github.com/w3c/wcag/blob/wcag22-focus-1038/guidelines/sc/22/focus-visible-enhanced.html) for changes based on this issue.

**Comment by AutoSponge** [comment]
@alastc I'm not taking issue with the language, I think it's clear. I'm taking issue with the rule. As I demonstrated, it's far too easy to "pass" by making a poorly visible focus indicator. A single 2px-wide line of an identical color to one of the adjacent colors and 3+:1 with the replaced color. By enforcing 3:1 with all adjacent colors, it's much harder to create a bad focus indicator.

<img width="347" alt="Screen Shot 2020-02-16 at 6 00 04 PM" src="https://user-images.githubusercontent.com/654530/74614483-3cbde700-50e6-11ea-9faa-fe1e97780541.png">

A minimum "fix" to my example would surround a white underline with blue instead of allowing either a white underline touching the white button or a blue underline touching the blue background.

**Comment by alastc** [comment]
Harder, but then you also prevent indicators which are very visible. I forked the codepen and added a list example:
https://codepen.io/alastc/pen/VwLaQMe

**Comment by AutoSponge** [comment]
@alastc You're right, and you used a different heuristic to make that more visible, changed area = 2*w + 2*h. Meaning, instead of 2px width being a deciding factor, the addition of all 4 sides having a 2px outline provide much better visibility. I think that would be a better predictive measure than the width (assuming it can be word-smithed into a "formula").

**Comment by JAWS-test** [comment]
According to @alastc [example](https://codepen.io/alastc/pen/VwLaQMe) it makes a great difference, 
* whether I add a thin line to a large area of the same color as a focus indicator (poorly visible)
* or whether I add the same thin line to a thin line of the same color as a focus indicator (clearly visible).

It would be good if the new criterion forbids the first variant and allows the second. However, it will be difficult to formulate this as a condition and unfortunately I have no idea how this could best be done. I am afraid that we cannot rule out the bad option unless we come up with a good formulation. At the very least, the Understanding could point out that certain focus indicators that formally meet the new criterion are poorly visible and should be avoided

**Comment by alastc** [comment]
> you used a different heuristic to make that more visible

Sure, I wasn't trying to create a different minimum example, I was trying to show that the blanket rule of "Must have adjacent contrast" creates a _false positive_. I.e. blocks a good indicator.

The aim is to create a formula that catches as many 'bad' cases as possible without catching false positives. 

> At the very least, the Understanding could point out that certain focus indicators that formally meet the new criterion are poorly visible and should be avoided

Sure, and we can confidently say that bigger & more contrasting are better. In fact, a 2px thick outline with sufficient "contrast change" is a safe bet. (I think, there's bound to be an exception!) 

Patrick's comments are making me think we lead with that exmple in the understanding doc.

But, for those wanting a different style/approach to focus indicators, I think it is useful to define this formula as best we can.

**Comment by alastc** [comment]
For anyone in the working group, there are some more proposals on the table:
https://www.w3.org/2002/09/wbs/35422/focus-visible-enh-issues1/

For anyone following along, the outline of those proposals is:
https://docs.google.com/document/d/1F83m-HXRkXz1QCF6_QNtQGzIULugMiFSLgtqDRgKA-s/edit#

**Comment by alastc** [comment]
So it looks like we are going with [option 3 in the document](https://docs.google.com/document/d/1F83m-HXRkXz1QCF6_QNtQGzIULugMiFSLgtqDRgKA-s/edit#heading=h.96ib39rkbmhb), which includes:

- Changes of the bullet names to "change of contrast" and "adjacent contrast".
- The note says "A keyboard focus indicator which has a pattern or gradient may have parts that do not meet the 3:1 contrast ratio for the change of contrast, as long as an area equal to the minimum does meet the contrast ratio."

Hopefully the other changes also help with the understandability?

**Comment by alastc** [comment]
Hi @patrickhlauke, given the changes in the [revised version](https://raw.githack.com/w3c/wcag/wcag22-focus-visible-enh-updates/understanding/22/focus-visible-enhanced.html), can this issue be closed now?

**Comment by scha14** [comment]
Hi @patrickhlauke, given [a few changes](https://raw.githack.com/w3c/wcag/wcag22-focus-visible-enh-updates/understanding/22/focus-appearance-minimum.html) in the new draft that address concerns above, closing this issue. Please reopen if there are unaddressed concerns. Thank you.
